File: src/main/java/me/aov/sellgui/listeners/ItemStackNormalizeListener.java
package me.aov.sellgui.listeners;

import me.aov.sellgui.SellGUIMain;
import me.aov.sellgui.utils.ItemStackNormalizer;
import org.bukkit.entity.Player;
import org.bukkit.event.EventHandler;
import org.bukkit.event.EventPriority;
import org.bukkit.event.Listener;
import org.bukkit.event.inventory.FurnaceExtractEvent;
import org.bukkit.event.inventory.FurnaceSmeltEvent;
import org.bukkit.event.player.PlayerJoinEvent;
import org.bukkit.event.player.PlayerQuitEvent;
import org.bukkit.inventory.ItemStack;

public class ItemStackNormalizeListener implements Listener {

    private final SellGUIMain plugin;

    public ItemStackNormalizeListener(SellGUIMain plugin) {
        this.plugin = plugin;
    }

    @EventHandler(priority = EventPriority.MONITOR)
    public void onPlayerJoin(PlayerJoinEvent event) {
        if (!isEnabled("stacking.normalize-on-join", true)) {
            return;
        }

        normalizeLater(event.getPlayer());
    }

    @EventHandler(priority = EventPriority.MONITOR)
    public void onPlayerQuit(PlayerQuitEvent event) {
        if (!isEnabled("stacking.normalize-on-quit", true)) {
            return;
        }

        ItemStackNormalizer.normalizePlayerInventory(plugin, event.getPlayer());
    }

    @EventHandler(priority = EventPriority.HIGHEST, ignoreCancelled = true)
    public void onFurnaceSmelt(FurnaceSmeltEvent event) {
        if (!isEnabled("stacking.normalize-smelt-results", true)) {
            return;
        }

        ItemStack result = event.getResult();
        if (result != null) {
            event.setResult(ItemStackNormalizer.normalizedCopy(plugin, result));
        }
    }

    @EventHandler(priority = EventPriority.MONITOR, ignoreCancelled = true)
    public void onFurnaceExtract(FurnaceExtractEvent event) {
        if (!isEnabled("stacking.normalize-after-furnace-extract", true)) {
            return;
        }

        normalizeLater(event.getPlayer());
    }

    private void normalizeLater(Player player) {
        plugin.getServer().getScheduler().runTaskLater(plugin,
                () -> ItemStackNormalizer.normalizePlayerInventory(plugin, player),
                1L);
    }

    private boolean isEnabled(String path, boolean defaultValue) {
        return plugin.getConfig().getBoolean("stacking.enabled", true)
                && plugin.getConfig().getBoolean(path, defaultValue);
    }
}

File: src/main/kotlin/me/aov/sellgui/listeners/ItemStackNormalizeListener.kt
package me.aov.sellgui.listeners

import me.aov.sellgui.SellGUIMain
import me.aov.sellgui.utils.ItemStackNormalizer
import org.bukkit.entity.Player
import org.bukkit.event.EventHandler
import org.bukkit.event.EventPriority
import org.bukkit.event.Listener
import org.bukkit.event.inventory.FurnaceExtractEvent
import org.bukkit.event.inventory.FurnaceSmeltEvent
import org.bukkit.event.player.PlayerJoinEvent
import org.bukkit.event.player.PlayerQuitEvent

class ItemStackNormalizeListener(private val plugin: SellGUIMain) : Listener {
    @EventHandler(priority = EventPriority.MONITOR)
    fun onPlayerJoin(event: PlayerJoinEvent) {
        if (isEnabled("stacking.normalize-on-join", true)) normalizeLater(event.player)
    }

    @EventHandler(priority = EventPriority.MONITOR)
    fun onPlayerQuit(event: PlayerQuitEvent) {
        if (isEnabled("stacking.normalize-on-quit", true)) ItemStackNormalizer.normalizePlayerInventory(plugin, event.player)
    }

    @EventHandler(priority = EventPriority.HIGHEST, ignoreCancelled = true)
    fun onFurnaceSmelt(event: FurnaceSmeltEvent) {
        event.result?.let { result ->
            if (isEnabled("stacking.normalize-smelt-results", true)) {
                ItemStackNormalizer.normalizedCopy(plugin, result)?.let { event.result = it }
            }
        }
    }

    @EventHandler(priority = EventPriority.MONITOR, ignoreCancelled = true)
    fun onFurnaceExtract(event: FurnaceExtractEvent) {
        if (isEnabled("stacking.normalize-after-furnace-extract", true)) normalizeLater(event.player)
    }

    private fun normalizeLater(player: Player) {
        plugin.server.scheduler.runTaskLater(plugin, Runnable { ItemStackNormalizer.normalizePlayerInventory(plugin, player) }, 1L)
    }

    private fun isEnabled(path: String, defaultValue: Boolean): Boolean =
        plugin.config.getBoolean("stacking.enabled", true) && plugin.config.getBoolean(path, defaultValue)
}
